src/olympia/zadmin/models.py

```diff
@@ -149,6 +149,58 @@ def apply_validation(self, validation):
         self.notices = results['notices'] + compat['notices']
         self.valid = self.errors == 0
 
+        messages = results['messages']
+        message_ids = ['.'.join(msg['id']) for msg in messages]
+
+        message_summary = {}
+
+        for message, message_id in zip(messages, message_ids):
+            compat_type = message.get('compatibility_type')
+
+            if not compat_type:
+                compat_type = message.get('type')
+
+            if message_id not in message_summary:
+                message_summary[message_id] = (
+                    ValidationResultMessage.objects.get_or_create(
+                        message_id=message_id,
+                        validation_job=self.validation_job,
+                        defaults={
+                            'message': message.get('message', ''),
+                            'compat_type': compat_type,
+                        }
+                    )[0]
+                )
+
+        for message in message_summary.values():
+            # All relevant messages are already created, to avoid
+            # any side-effects we run a separate UPDATE query instead
+            # of an explicit `.save()` so that the database handles atomic
+            # increments for us.
+            ValidationResultMessage.objects.filter(pk=message.pk).update(
+                addons_affected=F('addons_affected') + 1
+            )
+
+            ValidationResultAffectedAddon.objects.create(
+                message=message,
+                addon=self.file.version.addon)
+
+
+class ValidationResultMessage(ModelBase):
+    """Summary of validation result messages"""
+    validation_job = models.ForeignKey(ValidationJob,
+                                       related_name='message_summary')
+    message_id = models.CharField(max_length=256)
```

Comment:
It's confusing how `message_id` is a `CharField` here, but is a `ForeignKey` pointing to this `ValidationResultMessage` below, in `ValidationResultAffectedAddon`.